rust/noosphere-core/src/data/body_chunk.rs

```diff
@@ -70,4 +76,342 @@ impl BodyChunkIpld {
 
         Ok(all_bytes)
     }
+
+    /// Encode `content` as a [BodyChunkIpld] chain in streaming fashion.
+    pub async fn encode<R, S>(content: R, store: &S) -> Result<Cid>
+    where
+        R: AsyncRead + Unpin,
+        S: Scratch + BlockStore,
+    {
+        let mut chunker = AsyncStreamCDC::new(
+            content,
+            fastcdc::v2020::MINIMUM_MIN,
+            BODY_CHUNK_MAX_SIZE / 2,
+            BODY_CHUNK_MAX_SIZE,
+        );
+        let stream = chunker.as_stream();
+        tokio::pin!(stream);
+
+        let mut encoder = BodyChunkEncoder::new(store);
+        while let Some(chunk) = stream.try_next().await? {
+            encoder.push(chunk.data).await?;
+        }
```

Comment:
It would be nice if the encoder wrapped the stream directly and only poll the stream if the encoder itself were being streamed (for maximum laziness).

Separately but related: it would be nice to be able to get the `BodyChunkIpld`s produced by the encoder as a stream. Streams are nice because you can express both "I want to do something at each step" and "I want to do something after all steps" with the same primitive.